Shell/Options.cpp

```diff
@@ -3250,16 +3335,28 @@ bool Options::checkGlobalOptionConstraints(bool fail_early)
 
 /**
  * Check whether the option values make sense with respect to the given problem
+ *
+ * This check should be done at least twice; before preprocessing and after.
+ * With before_preprocessing on, only options tagged as PREPROCESSING are queried
+ * With before_preprocessing off, it's all the remaining ones.
+ *
  **/
-bool Options::checkProblemOptionConstraints(Property* prop,bool fail_early)
+bool Options::checkProblemOptionConstraints(Property* prop, bool before_preprocessing, bool fail_early)
 {
    CALL("Options::checkProblemOptionConstraints");
 
   bool result = true;
 
   VirtualIterator<AbstractOptionValue*> options = _lookup.values();
   while(options.hasNext()){
-    result = options.next()->checkProblemConstraints(prop) && result; 
+    AbstractOptionValue* opt = options.next();
+
+    if ((before_preprocessing && opt->getTag() != OptionTag::PREPROCESSING) ||
```

Comment:
```
bool tagIsPreprocessing = opt->getTag() == OptionTag::Preprocessing;
if(before_processing != tagIsPreprocessing)
```
? Just a matter of taste.

Reply:
It took me a while to get what you meant, but I agree your solution is easier to read in the end! ;)